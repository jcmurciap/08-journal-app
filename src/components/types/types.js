
export const types = {

    login: '[Auth] Login',
    logout: '[Auth] Logout',
    uiSetError: '[UI] set Error',
    uiRemoveError: '[UI] Remove Error',

    // blocking login botton
    uiStartLoading: '[UI] Start loading',
    uiFinishLoading: '[UI] Finish loading',
    notesAddNew: '[Notes] New note',
    notesActive: '[Notes] Set active note',
    notesLoad: '[Notes] Load notes',
    notesUpdated: '[Notes] Updated note',
    notesFileUrl: '[Notes] Updated image url',
    notesDelete: '[Notes] Delete note',
    notesLogoutCleaning: '[Notes] Logout Cleaning',   

}
