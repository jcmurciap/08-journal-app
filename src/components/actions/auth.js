// libraries
import Swal from 'sweetalert2';
import 'firebase/firestore';
import { getAuth, signInWithPopup, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword, signOut } from 'firebase/auth';
import { googleAuthProvider } from '../../firebase/firebase-config';

// components
import { finishLoading, startLoading } from './ui';
import { noteLogout } from './notes';
import { types } from '../types/types';

export const startLoginEmailPassword = ( email, password ) => {

    
    return ( dispatch ) => {
        // gateway to the Firebase authentication API
        const auth = getAuth();
        
        dispatch( startLoading() );
        
        return signInWithEmailAndPassword( auth, email, password )
            .then(({ user }) => {
                dispatch( login( user.uid, user.displayName ) );
                dispatch( finishLoading() );
            })   
            .catch( e => {
                dispatch( finishLoading() );
                Swal.fire( 'Error', e.message, 'error' );
            });
    }; 
}

// save data in firebase
export const startRegisterWithEmailPasswordName = ( name, email, password ) => {
    
    return ( dispatch ) => {
        const auth = getAuth();

        // autenticacion
        createUserWithEmailAndPassword( auth, email, password )
            .then( async({ user }) => {
                await updateProfile( user, { displayName: name } );
                dispatch( 
                    login( user.uid, user.displayName ) 
                );
            })
            .catch( e  => {
                Swal.fire( 'Error', e.message, 'error' );
            })
    }
}

export const startGoogleLogin = () => {
    return ( dispatch ) => {
        const auth = getAuth();
        signInWithPopup( auth, googleAuthProvider )
            .then( ({ user }) => {
                dispatch( login( user.uid, user.displayName ) );
            });
    };
}

export const login = ( uid, displayName ) => {
    return {
        type: types.login,
        payload: {
            uid,
            displayName 
        }
    };
}

export const startLogout = () => {    
    
    return async( dispatch ) => {
        
        const auth = getAuth();
        await signOut( auth );
        dispatch( logout() );
        dispatch( noteLogout() );  
    };
}

export const logout = () => ({
    type: types.logout
})
