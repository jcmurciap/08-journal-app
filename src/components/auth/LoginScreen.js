import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';

// components
import { useForm } from '../hooks/useForms';
import { startGoogleLogin, startLoginEmailPassword } from '../actions/auth';

export const LoginScreen = () => {
    
    const dispatch = useDispatch();
    
    const { loading } = useSelector( state => state.ui); // REVISE COMO FUNCIONA 
    
    const [ formValues, handleInputChange ] = useForm({}); 
    
    const { email, password } = formValues;
        
    const handleLogin = ( e ) => {
        e.preventDefault();
        dispatch( startLoginEmailPassword( email, password ) ); // registrado en redux
    };

    const handleGoogleLogin = () => {
        dispatch( startGoogleLogin() );
    };
    
    return (
        <>
            <div className='auth__container'>
                <h3 className="auth__title">
                    Sign In
                </h3>
                <form 
                    onSubmit={ handleLogin }
                    className="animate__animated animate__fadeIn animate__faster"
                >
                    <input 
                        autoComplete="off"
                        className="auth__input"
                        name="email"
                        onChange={ handleInputChange }    
                        placeholder="Email"
                        type="text"
                        value={ email }
                    />
                    <input 
                        className="auth__input"
                        name="password" 
                        onChange={ handleInputChange }  
                        placeholder="Password"
                        type="password"
                        value={ password } 
                    />
                
                    <button 
                        className="btn btn-primary btn-block"
                        disabled={ loading }
                        type="submit"
                    >
                        Login
                    </button>
                    
                    <div className="auth__social-networks">
                        <p>Login with social networks</p>
                        <div 
                            className="google-btn"
                            onClick={ handleGoogleLogin }
                        >
                            <div className="google-icon-wrapper">
                                <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                            </div>
                            <p className="btn-text">
                                <b>Sign In with google</b>
                            </p>
                        </div>
                    </div>
                    
                    <Link to="/auth/register" className="link">
                        Create new account
                    </Link>
                </form>
            </div>
        </>
    )
}
