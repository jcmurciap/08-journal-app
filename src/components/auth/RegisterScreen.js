import React from 'react';
import { Link } from 'react-router-dom';
import { useForm } from '../hooks/useForms';
import validator from 'validator';
import { removeError, setError } from '../actions/ui';
import { useDispatch, useSelector } from 'react-redux';
import { startRegisterWithEmailPasswordName } from '../actions/auth';

export const RegisterScreen = () => {
   
    const dispatch = useDispatch(); 
    const {  msgError } = useSelector( state => state.ui );
    
    const [ formValues, handleInputChange ] = useForm({
        fname: 'jose',
        email: 'jose@gmail.com',
        password: 'camilo',
        password2: 'camilo',
    });
    const { fname, email, password, password2 } = formValues;
    
    

    const handleRegister = (e) => {
        e.preventDefault();
        
        if ( isFormValid() ){

            dispatch( startRegisterWithEmailPasswordName( fname, email, password ) )
        }
    }

    const isFormValid = () => {
        if( fname.trim().length === 0 ){
            dispatch(setError('Name is requerided.\n\tTry again'));
            return false;
        }else if( !validator.isEmail( email ) ){
            dispatch(setError('Email is requerided.\n\tTry again'));
            return false;
        }else if( password !== password2 || password.length < 5 ) {
            dispatch(setError('Passwords needs to be equal and has at least 6 characters.\n\tTry again'));
            return false;
        }
        dispatch( removeError() );
        return true;
    }; 

    return (
        <>
            <h3 className="auth__title">
                Create New Account
            </h3>

            <form 
                onSubmit={ handleRegister }
                className="animate__animated animate__fadeIn animate__faster"
            >
                {
                    msgError && (
                        <div className="auth__alert-error">
                            { msgError }
                        </div>
                    )
                }
                <input 
                    type="text"
                    placeholder="Name"
                    name="fname"
                    className="auth__input"
                    autoComplete="off"
                    value={ fname }    
                    onChange={ handleInputChange }
                />
                <input 
                    type="text"
                    placeholder="Email"
                    name="email"
                    className="auth__input"
                    autoComplete="off"
                    value={ email }    
                    onChange={ handleInputChange }
                />
                <input 
                    type="password"
                    placeholder="Password"
                    name="password" 
                    className="auth__input"
                    value={ password }   
                    onChange={ handleInputChange }
                />
                <input 
                    type="password"
                    placeholder="Confirm Password"
                    name="password2" 
                    className="auth__input"
                    value={ password2 }   
                    onChange={ handleInputChange }
                />
            
                <button 
                    type="submit"
                    className="btn btn-primary btn-block mb-5"
                >
                    Register
                </button>
                
                <Link 
                    to="/auth/login"
                    className="link"
                >
                    Already registered?
                </Link>
            </form>
        </>
    )
}
