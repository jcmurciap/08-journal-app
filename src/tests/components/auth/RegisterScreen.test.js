import {mount} from 'enzyme';
import React from 'react';
import {Provider} from 'react-redux'
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import {MemoryRouter} from 'react-router-dom';
import { types } from '../../../components/types/types';

const middlewares = [thunk];

// create a store
const mockStore = configureStore(middlewares);

const initialState = {
    auth: {},
    ui: {
        loading: false,
        magError: null,
    },
    notes: {
        notes: [],
        active: null 
    },
};

let store = mockStore(initialState);

// mock the store
// store.dispatch = jest.fn(); 

describe('<RegisterScreen /> tests', () => {

    // reinicializacion del store para cada ejecuciÃ³n de una prueba
    // beforeEach( () => {
    //     store = mockStore( initialState );

    //     // good practice => restablish default values
    //     jest.clearAllMocks()
    // });
    
    const wrapper = mount(
        
        <Provider store={store}>
            <MemoryRouter>
                <RegisterScreen /> 
            </MemoryRouter>
        </Provider>
    );

    
    test( 'Should to return a skeleton', () => {

        expect(wrapper).toMatchSnapshot();

    });

    // vamos a cambiar el email para dejarlo vacio y evaluamos.
    test('make dispatch to respectively action', () => {

        // referencia al email (input)
        const emailField = wrapper.find('input[name="email"]');

        // simulate the change === onChange
        emailField.simulate('change', {
            // recuerde que estamos trabajando con el customHook 'useForm'
            target: {
                value: '',
                name: 'email',
            }
        });

        wrapper.find('form').simulate('submit', {
            preventDefault(){}
        });

        const actions = store.getActions();
        expect(actions[0]).toEqual({
            type: types.uiSetError,
            payload: 'Email is requerided.\n\tTry again'
        });
    });

    test( 'Show error box with alert message', () => {

        const initialState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'email has failed',
            },
        };

        const store = mockStore(initialState);

        const wrapper = mount(
        
            <Provider store={store}>
                <MemoryRouter>
                    <RegisterScreen /> 
                </MemoryRouter>
            </Provider>
        );
        
        expect(wrapper.find('.auth__alert-error').exists()).toBe(true);
        expect(wrapper.find('.auth__alert-error').text().trim()).toBe(initialState.ui.msgError);



    })








})











