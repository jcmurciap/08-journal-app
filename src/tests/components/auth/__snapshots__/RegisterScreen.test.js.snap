// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterScreen /> tests Should to return a skeleton 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Create New Account
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="fname"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value="jose"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="jose@gmail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="camilo"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirm Password"
      type="password"
      value="camilo"
    />
    <button
      className="btn btn-primary btn-block mb-5"
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
